File: 프로그래머스/1/250125. ［PCCE 기출문제］ 9번 ／ 이웃한 칸/［PCCE 기출문제］ 9번 ／ 이웃한 칸.py
def solution(board, h, w):
    answer = 0
    dh, dw = (-1, 0, 1, 0), (0, -1, 0, 1)
    
    for i in range(4):
        nh, nw = h + dh[i], w + dw[i]
        if nh < 0 or nh >= len(board) or nw < 0 or nw >= len(board[0]):
            continue
        if board[nh][nw] == board[h][w]:
            answer += 1
    
    return answer

File: 프로그래머스/1/250125. ［PCCE 기출문제］ 9번 ／ 이웃한 칸/［PCCE 기출문제］ 9번 ／ 이웃한 칸.java
class Solution {
    public int solution(String[][] board, int h, int w) {
        int count = 0;
        int n = board.length;
        int m = board[0].length;
        String color = board[h][w];

        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        for (int[] d : directions) {
            int nh = h + d[0];
            int nw = w + d[1];

            if (isValid(nh, nw, n, m) && board[nh][nw].equals(color)) {
                count++;
            }
        }

        return count;
    }

    private boolean isValid(int h, int w, int n, int m) {
        return h >= 0 && h < n && w >= 0 && w < m;
    }
}